client/cli/src/runner.rs

```diff
@@ -177,45 +179,61 @@ impl<C: SubstrateCli> Runner<C> {
 
 	/// A helper function that runs a future with tokio and stops if the process receives the signal
 	/// `SIGTERM` or `SIGINT`.
-	pub fn run_subcommand<BU, B, BA, IQ, CL>(self, subcommand: &Subcommand, builder: BU)
-		-> Result<()>
+	pub fn run_subcommand<B, FBA, IQ, FCL, FB, LB, LCL, LBA>(
+		self, subcommand: &Subcommand, full_builder: FB, light_builder: LB,
+	) -> Result<()>
 	where
-		BU: FnOnce(Configuration)
-			-> sc_service::error::Result<(Arc<CL>, Arc<BA>, IQ, TaskManager)>,
+		FB: FnOnce(Configuration)
+			-> sc_service::error::Result<(Arc<FCL>, Arc<FBA>, IQ, TaskManager)>,
+		LB: FnOnce(Configuration)
+			-> sc_service::error::Result<(Arc<LCL>, Arc<LBA>, NetworkStatusSinks<B>, TaskManager)>,
 		B: BlockT + for<'de> serde::Deserialize<'de>,
-		BA: sc_client_api::backend::Backend<B> + 'static,
+		FBA: sc_client_api::backend::Backend<B> + 'static,
 		IQ: sc_service::ImportQueue<B> + 'static,
 		<B as BlockT>::Hash: FromStr,
 		<<B as BlockT>::Hash as FromStr>::Err: Debug,
 		<<<B as BlockT>::Header as HeaderT>::Number as FromStr>::Err: Debug,
-		CL: UsageProvider<B> + BlockBackend<B> + StorageProvider<B, BA> + Send + Sync +
+		FCL: UsageProvider<B> + BlockBackend<B> + StorageProvider<B, FBA> + Send + Sync +
 		'static,
+		LCL: HeaderBackend<B>,
+		LBA: sc_service::chain_ops::MaybeChtRootStorageProvider<B>,
 	{
 		let chain_spec = self.config.chain_spec.cloned_box();
 		let network_config = self.config.network.clone();
 		let db_config = self.config.database.clone();
 
 		match subcommand {
-			Subcommand::BuildSpec(cmd) => cmd.run(chain_spec, network_config),
+			Subcommand::BuildSpec(cmd) => {
+				if cmd.export_sync_state {
+					let (client, backend, network_status_rx, task_manager) = light_builder(self.config)?;
+					run_until_exit(
+						self.tokio_runtime,
+						cmd.run_export_sync_state(chain_spec, network_config, client, backend, network_status_rx),
+						task_manager
+					)
```

Comment:
I know it's not directly related to this issue but I just noticed the run_until_exit is wrong.

Can you replace, in run_until_exit, this:
```
	task_manager.terminate();
	drop(tokio_runtime);
```

by this:

```
	task_manager.clean_shutdown();
```

Reply:
I occasionally get a few errors on shutdown after this such as
```
2020-08-18 10:24:29.218 tokio-runtime-worker WARN telemetry  Ignored telemetry message because of error on channel: TrySendError { kind: Disconnected }
```

```
Thread 'tokio-runtime-worker' panicked at 'slog-scope: No logger set. Use `slog_scope::set_global_logger` or `slog_scope::scope`.', /home/ashley/.cargo/registry/src/github.com-1ecc6299db9ec823/slog-scope-4.3.0/lib.rs:125
```

Reply:
It should improve, not the opposite XD

I will make a branch to see if that triggers error in the ci (the ci more easily triggered by wrong shutdown)

Reply:
#6909